Disable unsafe OpenSSL APIs/Macros (#3492)
openenclave/openenclave: docs/OpenSSLSupport.md

but also other OpenSSL APIs that internally depend on the RAND APIs. Alternatively, developers
can implement their own RAND method to replace the default method via `RAND_set_rand_method` API.

> ## Security Guidance of using OpenSSL APIs/Macros

Reviewer: ```suggestion
## Security Guidance for using OpenSSL APIs/Macros
```
Note: assuming you accept this suggestion, you will need to update the cross-reference links to this section below.

> Few OpenSSL APIs allow users to configure an application to load files from the host filesystem.
> Such behaviors may have security implications given that the host filesystem is untrusted under the
> threat model of OE. For these reasons, OE recommends users not to use those APIs and
> put them into `deprecations.h` as part of OpenSSL headers that are shipped along with the release package.
> OE's default compiler settings explicitly include the header such that users get warnings when an
> enclave uses any of these APIs.

Reviewer 2: ```suggestion
OpenSSL provides APIs that allow users to configure sensitive settings like certificate trust and cryptosuite preference from files.
Because the host file system is considered untrusted in contexts such as SGX enclaves, OE SDK marks these APIs as deprecated to discourage their use.
OE SDK does this by including a "deprecations.h" file as part of the default cmake/pkgconfig build configuration,
which will produce compiler warnings when any of these APIs are used in an enclave.
The warnings can be disabled by specifying the `OE_DISABLE_OPENSSL_DEPRECATIONS` option to the compiler.
```

See the following table for the detailed list of APIs.

API | Original header | Comments
:---:|:---:|:---|

> SSL_CTX_load_verify_locations | ssl.h | The API specifies the default locations from which an application looks up CA certificates for verification purposes. The API would allow an untrusted host to force the application to load fake certificates. |

Reviewer 2: We should probably also provide guidance as to what the alternative that a developer should use is.

> SSL_CTX_load_verify_locations | ssl.h | The API specifies the default locations from which an application looks up CA certificates for verification purposes. The API would allow an untrusted host to force the application to load fake certificates. |

Author: Updated the comments to be more accurate.

SSL_CTX_set_default_verify_paths | ssl.h | The API specifies the locations as part of arguments from which an application looks up CA certificates for verification purposes. The API would allow an untrusted host to force the application to load fake certificates. |
SSL_CTX_set_default_verify_dir | ssl.h | Similar to SSL_CTX_set_default_verify_paths except for specifying dir only. |
SSL_CTX_set_default_verify_file | ssl.h | Similar to SSL_CTX_set_default_verify_paths except for specifying file only. |

> X509_LOOKUP_hash_dir | x509_vfy.h | This API returns a `X509_LOOKUP_METHOD` method that loads files from the path specified by the `SSL_CERT_DIR` environment variable. The API is used by SSL_CTX_set_default_verify_dir and SSL_CTX_set_default_verify_paths internally.  |

Author: I added a recommendation to implement a customized `X509_LOOPUP_METHOD`. Will that be sufficient?

> X509_LOOKUP_hash_dir | x509_vfy.h | This API returns a `X509_LOOKUP_METHOD` method that loads files from the path specified by the `SSL_CERT_DIR` environment variable. The API is used by SSL_CTX_set_default_verify_dir and SSL_CTX_set_default_verify_paths internally. | The recommendation is implementing a customized `X509_LOOKUP_METHOD` method based on in-memory operations. |

Reviewer 2: >> Might be useful to clarify that the X509_LOOKUP methods are not blocked because they are used SSL_CTX_set_default_verify_* methods; they're blocked because they provide analogous functionality to X509_load_*_file methods.
>
>I added a recommendation to implement a customized X509_LOOPUP_METHOD. Will that be sufficient?

My previous comment here was with regards to the Comment column, not the newly added Guidance column; i.e. the Comments should explain why the API is blocked (it provides way to load trust certificates and CRLs from untrusted host) rather than functional details (it is used in the SSL_CTX_* methods). 

Regarding the guidance, I think that's fine, though I would suggest adding a bit more context to the recommendation. For example:
```md
This method is used internally by `SSL_CTX_set_default_verify_dir` and `SSL_CTX_set_default_verify_paths`.
You can define your own enclave-specific [`X509_LOOKUP_METHOD`](
https://www.openssl.org/docs/man1.1.1/man3/X509_LOOKUP_meth_new.html)
but it will not be used by the SSL_CTX methods transparently. We recommend
explicitly setting certificate trust store witch `SSL_CTX_set_cert_store` instead.
```

Author: Updated.

Author: Note: the cross reference does not seem to work in the table.

> X509_LOOKUP_file | x509_vfy.h | This API returns a `X509_LOOKUP_METHOD` method that loads files from the path specified by the `SSL_CERT_FILE` environment variable. The API is used by SSL_CTX_set_default_verify_file and SSL_CTX_set_default_verify_paths internally. | The recommendation is not to use this API (i.e., the lookup method associated with it). | Same recommendation as X509_LOOKUP_hash_dir. |

Reviewer 2: This row has an extra column, so the Guidance ends up displaying as

> The recommendation is not to use this API (i.e., the lookup method associated with it).

Instead of the intended:

> Same recommendation as X509_LOOKUP_hash_dir.

Also, for this and `X509_STORE_set_default_paths`, I would suggest just copy-pasting the guidance rather than having the reader go hunt in the table for the reference.

Author: Thanks for the catch!

sha.h | Yes | - |
srp.h | No | SRC is disabled by OE. |
srtp.h | Yes | - |

> ssl.h | Partial | SSL2 and SSL3 methods are disabled. Heartbeats extension is disabled by default. Functions that are deprecated by OE for security concerns include: `SSL_CTX_set_default_verify_paths`, `SSL_CTX_set_default_verify_dir`, `SSL_CTX_set_default_verify_file`, `SSL_CTX_load_verify_locations`. Refer to [Security Guidance of using OpenSSL APIs/Macros](#security-guidance-of-using-openssl-apismacros) for more detail |

Reviewer: ```suggestion
ssl.h | Partial | SSL2 and SSL3 methods are disabled. Heartbeats extension is disabled by default. Functions that are unsupported by OE for security concerns include: `SSL_CTX_set_default_verify_paths`, `SSL_CTX_set_default_verify_dir`, `SSL_CTX_set_default_verify_file`, `SSL_CTX_load_verify_locations`. Refer to [Security Guidance of using OpenSSL APIs/Macros](#security-guidance-of-using-openssl-apismacros) for more detail |
```

uierr.h | Yes | - |
whrlpool.h | No | Whirlpool is disabled by OE. |
x509.h | Yes | - |

> x509_vfy.h | Partial | Functions that are deprecated by OE for security concerns include: `X509_load_cert_file`, `X509_load_crl_file`, `X509_LOOKUP_hash_dir`, `X509_LOOKUP_file`, `X509_load_cert_crl_file`, `X509_STORE_load_locations`, `X509_STORE_set_default_paths`. Refer to [Security Guidance of using OpenSSL APIs/Macros](#security-guidance-of-using-openssl-apismacros) for more detail. |

Reviewer: ```suggestion
x509_vfy.h | Partial | Functions that are unsupported by OE for security concerns include: `X509_load_cert_file`, `X509_load_crl_file`, `X509_LOOKUP_hash_dir`, `X509_LOOKUP_file`, `X509_load_cert_crl_file`, `X509_STORE_load_locations`, `X509_STORE_set_default_paths`. Refer to [Security Guidance of using OpenSSL APIs/Macros](#security-guidance-of-using-openssl-apismacros) for more detail. |
```
